internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Store implements an in-memory resource database using go-memdb.
+//
+// It can be used as a storage backend directly via the Backend type in this
+// package, but also handles reads in our Raft backend, and can be used as a
+// local cache when storing data in external systems (e.g. RDBMS, K/V stores).
+type Store struct {
+	db  *memdb.MemDB
+	pub *stream.EventPublisher
+}
+
+// NewStore creates a Store.
+//
+// You must call Run before using the store.
+func NewStore() (*Store, error) {
+	db, err := memdb.NewMemDB(&memdb.DBSchema{
+		Tables: map[string]*memdb.TableSchema{
+			tableNameMetadata: {
+				Name: tableNameMetadata,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      &memdb.StringFieldIndex{Field: "Key"},
+					},
+				},
+			},
+			tableNameResources: {
+				Name: tableNameResources,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      idIndexer{},
+					},
+					indexNameOwner: {
+						Name:         indexNameOwner,
+						AllowMissing: true,
+						Unique:       false,
+						Indexer:      ownerIndexer{},
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Store{
+		db:  db,
+		pub: stream.NewEventPublisher(10 * time.Second),
+	}
+	s.pub.RegisterHandler(eventTopic, s.watchSnapshot, false)
+
+	return s, nil
+}
+
+// Run until the given context is canceled. This method blocks, so should be
+// called in a goroutine.
+func (s *Store) Run(ctx context.Context) { s.pub.Run(ctx) }
+
+// Read a resource using its ID.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) Read(id *pbresource.ID) (*pbresource.Resource, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	val, err := tx.First(tableNameResources, indexNameID, id)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return nil, storage.ErrNotFound
+	}
+
+	res := val.(*pbresource.Resource)
+
+	// Observe the Uid if it was given.
+	if id.Uid != "" && res.Id.Uid != id.Uid {
+		return nil, storage.ErrNotFound
+	}
+
+	// Let the caller know they need to upgrade/downgrade the schema version.
+	if id.Type.GroupVersion != res.Id.Type.GroupVersion {
+		return nil, storage.GroupVersionMismatchError{
+			RequestedType: id.Type,
+			Stored:        res,
+		}
+	}
+
+	return res, nil
+}
+
+// WriteCAS performs an atomic Check-And-Set (CAS) write of a resource.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) WriteCAS(res *pbresource.Resource, vsn string) error {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	existing, err := tx.First(tableNameResources, indexNameID, res.Id)
+	if err != nil {
+		return err
+	}
+
+	if existing == nil && vsn != "" {
```

Comment:
nit: Maybe add a comment about all creations using the WriteCAS method are expected to be issued with an empty version. It took me a moment to figure out that this was correct in order to prevent two creations of the same resource racing in odd ways.

Reply:
Ah yes, good call. It's non-obvious for sure!